### PR TITLE
added shellcheck linting step to make verify

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Adds a linting step for our Bash files so that side-effects can hopefully be reduced.

@deads2k @liggitt PTAL

This will need `vargrant-openshift` change https://github.com/openshift/vagrant-openshift/pull/425

### CODE REVIEW

**Comment by deads2k** [comment]
How about choosing one file and demonstrating what the linter picks up?

**Comment by stevekuznetsov** [comment]
The list of things that it can pick up is in the [README](https://github.com/koalaman/shellcheck#gallery-of-bad-code). Our files are *rife* with errors. There's ten-to-twenty in every file.

**Comment by deads2k** [comment]
Let's pretend I want a visual example in `./test/cmd/admin.sh`

**Comment by stevekuznetsov** [comment]
```
$ shellcheck test/cmd/admin.sh 

In test/cmd/admin.sh line 37:
os::cmd::expect_failure_and_text 'openshift start network --config=${NODECONFIG} --enable=kubelet' 'the following components are not recognized: kubelet'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In test/cmd/admin.sh line 38:
os::cmd::expect_failure_and_text 'openshift start network --config=${NODECONFIG} --enable=kubelet,other' 'the following components are not recognized: kubelet, other'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In test/cmd/admin.sh line 39:
os::cmd::expect_failure_and_text 'openshift start network --config=${NODECONFIG} --disable=other' 'the following components are not recognized: other'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In test/cmd/admin.sh line 40:
os::cmd::expect_failure_and_text 'openshift start network --config=${NODECONFIG} --disable=proxy,plugins' 'at least one node component must be enabled \(plugins, proxy\)'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In test/cmd/admin.sh line 42:
os::cmd::expect_failure_and_text 'openshift start node --config=${NODECONFIG} --disable=other' 'the following components are not recognized: other'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In test/cmd/admin.sh line 43:
os::cmd::expect_failure_and_text 'openshift start node --config=${NODECONFIG} --disable=kubelet,proxy,plugins' 'at least one node component must be enabled \(kubelet, plugins, proxy\)'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In test/cmd/admin.sh line 253:
os::util::sed "s/5000/6000/g" ${workingdir}/oadm_registry.yaml
                              ^-- SC2086: Double quote to prevent globbing and word splitting.


In test/cmd/admin.sh line 358:
os::cmd::expect_success_and_text 'oc get user/~ --token="$( oc sa get-token my-sa-name )"' 'system:serviceaccount:.+:my-sa-name'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.


In test/cmd/admin.sh line 361:
os::cmd::expect_success_and_text 'oc get user/~ --token="$( oc sa new-token my-sa-name )"' 'system:serviceaccount:.+:my-sa-name'
                                 ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.
```

**Comment by deads2k** [comment]
Didn't you specifically suggest that we use single quotes in every one of those cases?

**Comment by stevekuznetsov** [comment]
There were two angles for that:
 - letting the variables expand before `os::cmd` allows the output of `os::cmd` to be readable - often the reason something failed was a mis-expansion when expansions are present, and regardless if you want to re-create the problem yourself you often want to know what the expansion was
 - some expansions are not stable and expanding them will cause difficulties in tracking for jUnit as the "title" of the test is therefore unstable - I think this has some credence but honestly with the amount of change going on unless we create a framework where we can name test cases this point is moot

So no, in this case I'd prefer things to be expanded.

**Comment by stevekuznetsov** [comment]
And in either case, we can disable specific ShellCheck checks per file, or per line, etc. I think that the *necessity* of doing explicit disables is a boon. If you are intentionally not quoting a variable because you *want* word splitting, for instance, that's something that should be noted in a comment so it is obvious. By necessitating that a ShellCheck ignore statement be there for CI to pass you're making the cost of placing an explanatory comment that much smaller.

**Comment by deads2k** [comment]
So I'd specify a line number that is unstable?  Non-starter.  Disabling a given check per-file dilutes the utility.  Is there a third option that would actually work?

**Comment by stevekuznetsov** [comment]
No, not by line number. The disable is a comment directly adjacent to the line being disabled.

**Comment by deads2k** [comment]
> No, not by line number. The disable is a comment directly adjacent to the line being disabled.

Alright, can you find some examples that are valuable?  Your snippet really isn't.

**Comment by stevekuznetsov** [comment]
Added example commit for `hack/util.sh` Original errors were: https://gist.github.com/stevekuznetsov/a14e6c7e63a3908a14ff

**Comment by stevekuznetsov** [comment]
@Miciah please take a look at the last commit

File: hack/util.sh [hunk 1]
```diff
@@ -131,14 +131,14 @@ function start_os_server {
 	echo "[INFO] Using images:              ${USE_IMAGES}"
 	echo "[INFO] MasterIP is:               ${MASTER_ADDR}"
 
-	mkdir -p ${LOG_DIR}
+	mkdir -p "${LOG_DIR}"
 
 	echo "[INFO] Scan of OpenShift related processes already up via ps -ef	| grep openshift : "
-	ps -ef | grep openshift
+	pgrep -ef openshift
```
Comment (Miciah) [inline hack/util.sh]: What does `-e` to `pgrep` do?

File: hack/util.sh [hunk 2]
```diff
@@ -131,14 +131,14 @@ function start_os_server {
 	echo "[INFO] Using images:              ${USE_IMAGES}"
 	echo "[INFO] MasterIP is:               ${MASTER_ADDR}"
 
-	mkdir -p ${LOG_DIR}
+	mkdir -p "${LOG_DIR}"
 
 	echo "[INFO] Scan of OpenShift related processes already up via ps -ef	| grep openshift : "
```
Comment (Miciah) [inline hack/util.sh]: This `echo` statement becomes out of synch with the actual command run.

Comment (Miciah) [inline hack/util.sh]: Also, the output of `ps` might be more useful than the list of pids from `pgrep`, so I don't know whether you really want to change it.

File: hack/util.sh [hunk 3]
```diff
@@ -228,7 +228,8 @@ function ensure_iptables_or_die {
 # tryuntil loops, retrying an action until it succeeds or times out after 90 seconds.
 function tryuntil {
 	timeout=$(($(date +%s) + 90))
-	echo "++ Retrying until success or timeout: ${@}"
+	echo "++ Retrying until success or timeout: $*"
+	# shellcheck disable=SC2078
 	while [ 1 ]; do
```
Comment (Miciah) [inline hack/util.sh]: Would using a more shell-idiomatic `while :` or `while true` suppress the warning?

File: hack/util.sh [hunk 4]
```diff
@@ -261,21 +262,25 @@ function wait_for_command {
...
 		return 0
 	fi
 	#check a failure condition where the success
 	#command may never be evaulated before timing
 	#out
 	if [[ ! -z $fail ]]; then
-		eval $fail
+		# we want word splitting on the call to `eval`
+		# shellcheck disable=SC2086
+		eval ${fail} # @miciah: can we get rid of `eval` here?
```
Comment (Miciah) [inline hack/util.sh]: Alas, we sometimes give `wait_for_command` commands that use pipes, and it seems Bash handles redirection *after* expansion, so we do need to use `eval`.  However, I don't see why we need word-splitting before `eval` runs the given command (it'll perform word-splitting when it evaluates the command).

File: hack/util.sh (hunk 2 above)
Comment (stevekuznetsov) [inline hack/util.sh]: I'm not sure that the warning about `pgrep` over `ps | grep` is useful, so I blacklisted it.

**Comment by stevekuznetsov** [comment]
[test] me

**Comment by stevekuznetsov** [comment]
@Miciah addressed comments

File: hack/util.sh [hunk 5]
```diff
@@ -301,15 +306,15 @@ function wait_for_url_timed {
...
-	expire=$(($(time_now) + $max_wait))
+	expire=$(($(time_now) + max_wait))
 	set +e
 	while [[ $(time_now) -lt $expire ]]; do
-	out=$(curl --max-time 2 -fs $url 2>/dev/null)
+	out=$(curl --max-time 2 -fs "$url" 2>/dev/null)
 	if [ $? -eq 0 ]; then
 		set -e
-		echo ${prefix}${out}
+		echo "${prefix}${out}"
```
Comment (Miciah) [inline hack/util.sh]: This change means we won't clean up whitespace (i.e., any newlines in `$out` will be retained, and consecutive blanks will no longer be squeezed), but AFAICS the output of `wait_for_url_timed` is intended only for humans to read anyway, so it shouldn't break anything.

Comment (stevekuznetsov) [inline hack/util.sh]: Right, this line is to be blamed for most of the spam that happens whenever we start the server using these scripts...

File: hack/util.sh [hunk 6]
```diff
@@ -351,21 +356,21 @@ function wait_for_url {
...
 			set -e
 			echo "${prefix}${out}"
 			return 0
 		fi
-		sleep $wait
+		sleep "${wait}"
 	done
 	echo "ERROR: gave up waiting for ${url}"
-	echo $(${cmd})
+	${cmd}
```
Comment (Miciah) [inline hack/util.sh]: You will be missing a blank line, if that's what's expected, in cases where `${cmd}` prints no output.  It's probably OK though.

File: hack/util.sh [hunk 7]
```diff
@@ -443,14 +448,14 @@ reset_tmp_dir() {
 	local sudo="${USE_SUDO:+sudo}"
 
 	set +e
-	${sudo} rm -rf ${BASETMPDIR} &>/dev/null
+	${sudo} rm -rf "${BASETMPDIR}" &>/dev/null
 	if [[ $? != 0 ]]; then
 		echo "[INFO] Unmounting previously used volumes ..."
-		findmnt -lo TARGET | grep ${BASETMPDIR} | xargs -r ${sudo} umount
-		${sudo} rm -rf ${BASETMPDIR}
+		findmnt -lo TARGET | grep "${BASETMPDIR}" | xargs -r "${sudo}" umount
```
Comment (Miciah) [inline hack/util.sh]: Quoting `${sudo}` here means it will expand to an empty parameter if it is empty, which will cause `xargs` to fail:

```console
$ empty=
$ echo foo | xargs -r $empty /bin/echo
foo
$ echo foo | xargs -r "$empty" /bin/echo
xargs: : No such file or directory
$
```

File: hack/util.sh [hunk 8]
```diff
@@ -460,17 +465,17 @@ function kill_all_processes()
 {
 	local sudo="${USE_SUDO:+sudo}"
 
-	pids=($(jobs -pr))
-	for i in ${pids[@]-}; do
-		pgrep -P "${i}" | xargs $sudo kill &> /dev/null
-		$sudo kill ${i} &> /dev/null
+	pids="$(jobs -pr)"
+	for i in ${pids}; do
+		pgrep -P "${i}" | xargs "${sudo}" kill &> /dev/null
```
Comment (Miciah) [inline hack/util.sh]: Same issue with `xargs` as above if `$sudo` is empty.

File: hack/util.sh [hunk 9]
```diff
@@ -536,13 +541,13 @@ function cleanup_openshift {
...
-		oc login -u system:admin -n default --config=${ADMIN_KUBECONFIG}
-		oc export all --all-namespaces --raw -o json --config=${ADMIN_KUBECONFIG} > ${LOG_DIR}/export_all.json
+		oc login -u system:admin -n default --config="${ADMIN_KUBECONFIG}"
+		oc export all --all-namespaces --raw -o json --config="${ADMIN_KUBECONFIG}" > "${LOG_DIR}/export_all.json"
 	fi
 
 	echo "[INFO] Dumping etcd contents to ${ARTIFACT_DIR}/etcd_dump.json"
 	set_curl_args 0 1
-	curl -s ${clientcert_args} -L "${API_SCHEME}://${API_HOST}:${ETCD_PORT}/v2/keys/?recursive=true" > "${ARTIFACT_DIR}/etcd_dump.json"
+	curl -s "${clientcert_args}" -L "${API_SCHEME}://${API_HOST}:${ETCD_PORT}/v2/keys/?recursive=true" > "${ARTIFACT_DIR}/etcd_dump.json"
```
Comment (Miciah) [inline hack/util.sh]: Don't we want word splitting on `$clientcert_args`?

File: hack/util.sh (hunk 7 above)
Comment (stevekuznetsov) [inline hack/util.sh]: Do you think we should turn off the warning here or not use `xargs`? Should we make a point of not using `xargs` and instead:
```sh
for mount in $( findmnt -lo TARGET | grep "${BASETMPDIR}" ); do
    ${sudo} umount "${mount}"
done
```
or something similar for read-ability?

Comment (stevekuznetsov) [inline hack/util.sh]: That change might actually make this method better -- currently it's written as if someone were trying to remove the tmpdirs themselves, and on noticing that the `rm -rf` failed, they `umount` and try again - with the `for` loop we wouldn't need the `if`, we could just `umount` any that we find and remove afterword.

File: hack/util.sh (hunk 8 above)
Comment (stevekuznetsov) [inline hack/util.sh]: Here likewise we don't really need the `xargs` if we just loop over PIDs in the first place

**Comment by stevekuznetsov** [comment]
@Miciah some questions on the use of `xargs`

File: hack/util.sh [hunk 10]
```diff
@@ -460,17 +458,18 @@ function kill_all_processes()
 {
 	local sudo="${USE_SUDO:+sudo}"
 
-	pids=($(jobs -pr))
-	for i in ${pids[@]-}; do
-		pgrep -P "${i}" | xargs $sudo kill &> /dev/null
-		$sudo kill ${i} &> /dev/null
+	for parent in $(jobs -pr); do
+		for child in $(pgrep -P "${parent}"); do
+			"${sudo}" kill "${child}" &> /dev/null
```
Comment (stevekuznetsov) [inline hack/util.sh]: @Miciah is this really the best approach here? I remember when I was doing my refactors (that have yet to land ... sometime soon!) I found that sometimes killing the child will corrupt the parent, and sometimes the parent has a 'keepalive' type of relationship with it's children.

Should we instead kill the parent first, and then later try killing the children if they're still alive?

Should we be explicitly using `kill -SIGTERM` here, and a follow-up `kill -SIGKILL` after some timeout? I believe I do that in my refactor so maybe most of these changes should just wait until that.

File: hack/util.sh [hunk 11]
```diff
@@ -593,16 +598,16 @@ function install_router {
...
-                 --cert=${MASTER_CONFIG_DIR}/router.crt --key=${MASTER_CONFIG_DIR}/router.key
-            cat ${MASTER_CONFIG_DIR}/router.crt ${MASTER_CONFIG_DIR}/router.key \
-                ${MASTER_CONFIG_DIR}/ca.crt > ${MASTER_CONFIG_DIR}/router.pem
+                 --cert="${MASTER_CONFIG_DIR}/router.crt" --key="${MASTER_CONFIG_DIR}/router.key"
+            cat "${MASTER_CONFIG_DIR}/router.crt" "${MASTER_CONFIG_DIR}/router.key" \
+                "${MASTER_CONFIG_DIR}/ca.crt" > "${MASTER_CONFIG_DIR}/router.pem"
             ROUTER_DEFAULT_CERT="--default-cert=${MASTER_CONFIG_DIR}/router.pem"
         fi
-        openshift admin router --create --credentials="${MASTER_CONFIG_DIR}/openshift-router.kubeconfig" --config="${ADMIN_KUBECONFIG}" --images="${USE_IMAGES}" --service-account=router ${ROUTER_DEFAULT_CERT-}
+        openshift admin router --create --credentials="${MASTER_CONFIG_DIR}/openshift-router.kubeconfig" --config="${ADMIN_KUBECONFIG}" --images="${USE_IMAGES}" --service-account=router "${ROUTER_DEFAULT_CERT-}"
```
Comment (Miciah) [inline hack/util.sh]: If `$ROUTER_DEFAULT_CERT` is empty, do you want to give the command an empty paramater?

File: hack/util.sh [hunk 12]
```diff
@@ -732,7 +739,7 @@ os::log::error_exit() {
 
 	local source_file=${BASH_SOURCE[$stack_skip]}
 	local source_line=${BASH_LINENO[$((stack_skip - 1))]}
-	echo "!!! Error in ${source_file}:${source_line}" >&2
+	echo "!!! Error in ${source_file}:${source_line}: ${message}" >&2
```
Comment (Miciah) [inline hack/util.sh]: Unrelated change.

File: hack/util.sh [hunk 13]
```diff
@@ -804,7 +811,7 @@ os::util::get_object_assert() {
   local request=$2
   local expected=$3
 
-  res=$(eval oc get $object -o go-template=\"$request\")
+  res=$(eval oc get "${object}" -o go-template="\"${request}\"")
```
Comment (Miciah) [inline hack/util.sh]: Sidenote: Doesn't look like this function is used anywhere.

File: hack/util.sh (hunk 12 above)
Comment (stevekuznetsov) [inline hack/util.sh]: `$message` was unused, so I used it.

File: hack/util.sh [hunk 14]
```diff
@@ -435,21 +435,19 @@ function validate_response {
 
 
 # reset_tmp_dir will try to delete the testing directory.
-# If it fails will unmount all the mounts associated with
+# If necesary, it will unmount all the mounts associated with
```
Comment (Miciah) [inline hack/util.sh]: s/necesary/necessary/

File: hack/util.sh (hunk 13 above)
Comment (stevekuznetsov) [inline hack/util.sh]: That's a good thing - I'm working to remove this and `tryuntil` as soon as I can.

File: hack/util.sh (hunk 11 above)
Comment (stevekuznetsov) [inline hack/util.sh]: Interesting, I'm not seeing where `openshift admin router` accepts a cert as an arg [here](https://github.com/openshift/origin/blob/master/pkg/cmd/admin/router/router.go#L433-L441).

@smarterclayton thoughts?

Comment (stevekuznetsov) [inline hack/util.sh]: Ah nevermind the flag is inside of the envar.

File: hack/util.sh [hunk 15]
```diff
@@ -460,10 +458,11 @@ function kill_all_processes()
 {
 	local sudo="${USE_SUDO:+sudo}"
 
-	pids="$(jobs -pr)"
-	for i in ${pids}; do
-		pgrep -P "${i}" | xargs "${sudo}" kill &> /dev/null
-		$sudo kill "${i}" &> /dev/null
+	for parent in $(jobs -pr); do
+		for child in $(pgrep -P "${parent}"); do
+			"${sudo}" kill "${child}" &> /dev/null
```
Comment (Miciah) [inline hack/util.sh]: You will still hit the issue with quoting `$sudo` here if it is empty.

Comment (stevekuznetsov) [inline hack/util.sh]: Good catch, didn't mean to have it there.

File: hack/verify-shellcheck.sh [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+./contrib/vagrant/provision-minimal.sh
+./contrib/vagrant/provision-full.sh
+./contrib/vagrant/provision-dind.sh
+./contrib/node/install-sdn.sh
+./tools/junitreport/test/integration.sh
+'
+for file in ${files}; do
+    if ! echo "${blacklist}" | grep -q "${file}"; then
+        if ! shellcheck "${file}"; then
+            failed=true
```
Comment (Miciah) [inline hack/verify-shellcheck.sh]: Probably should initialize `failed=false` outside the loop.

File: hack/verify-shellcheck.sh [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+./contrib/vagrant/provision-util.sh
+./contrib/vagrant/provision-master.sh
+./contrib/vagrant/provision-minimal.sh
+./contrib/vagrant/provision-full.sh
+./contrib/vagrant/provision-dind.sh
+./contrib/node/install-sdn.sh
+./tools/junitreport/test/integration.sh
+'
+for file in ${files}; do
+    if ! echo "${blacklist}" | grep -q "${file}"; then
```
Comment (Miciah) [inline hack/verify-shellcheck.sh]: It might be premature optimisation, but I'd definitely use an associative array rather than a `grep` for every file.

File: hack/verify-shellcheck.sh (hunk 16 above)
Comment (stevekuznetsov) [inline hack/verify-shellcheck.sh]: The check uses a default on unset, so we don't need to - were you thinking to do it that way for legibility?

File: hack/verify-shellcheck.sh (hunk 17 above)
Comment (stevekuznetsov) [inline hack/verify-shellcheck.sh]: OSX doesn't ship with associative arrays, unfortunately.

File: hack/util.sh (hunk 10 above)
Comment (Miciah) [inline hack/util.sh]: Sure, you could just append all the pids to a list and `kill -TERM` them all together, sleep, then repeat with `kill -KILL`.

File: hack/util.sh [hunk 18]
```diff
@@ -609,9 +617,9 @@ function install_router {
                  --cert="${MASTER_CONFIG_DIR}/router.crt" --key="${MASTER_CONFIG_DIR}/router.key"
             cat "${MASTER_CONFIG_DIR}/router.crt" "${MASTER_CONFIG_DIR}/router.key" \
                 "${MASTER_CONFIG_DIR}/ca.crt" > "${MASTER_CONFIG_DIR}/router.pem"
-            ROUTER_DEFAULT_CERT="--default-cert=${MASTER_CONFIG_DIR}/router.pem"
+            router_args+=("--default-cert='${MASTER_CONFIG_DIR}/router.pem'")
```
Comment (Miciah) [inline hack/util.sh]: You're overquoting.  This adds an element with literal quotation marks, and when you use `${router_args[*]}` below, you'll pass the quotation marks to `openshift admin router`.  See the following example:

```console
$ foo=( --foo --bar ); foo+=( "--baz='quux'" ); echo args: ${foo[*]}
args: --foo --bar --baz='quux'
```

Does `openshift admin router` strip quotation marks from the `--default-cert` flag value?

File: hack/verify-shellcheck.sh (hunk 16 above)
Comment (Miciah) [inline hack/verify-shellcheck.sh]: Legibility, or style—the principle that variables should be initialised before use.

Another thought: What about the admittedly unlikely situation that the user happens to have set `failed` in the environment before running `make verify`?

File: hack/verify-shellcheck.sh (hunk 17 above)
Comment (Miciah) [inline hack/verify-shellcheck.sh]: Oh.  That makes me sad.

Another idea is to put the blacklist into a file `blacklist.txt` and use `grep -f`.  Or use `comm`, but that's probably too advanced if the performance cost of `grep` in a `for` loop is negligible.

File: hack/util.sh (hunk 18 above)
Comment (stevekuznetsov) [inline hack/util.sh]: I was intending this - don't we want the quotation marks so that the flag value gets parsed as one item in the case that `$MASTER_CONFIG_DIR` has whitespace in it?

Comment (Miciah) [inline hack/util.sh]: You want to quote to avoid shell word-splitting, but you don't want to pass literal quotes to the command AFAIK.  What you really need is to omit the single-quotes here and use `"${router_args[@]}"` below:

```console
$ foo=( --foo --bar ); foo+=( "--baz='quux xyzzy'" )
$ printf ':%s: ' ${foo[*]}; echo
:--foo: :--bar: :--baz='quux: :xyzzy':
```

It's passing the quote marks to the command *and* splitting `--baz='quux xyzzy'` .

```console
$ printf ':%s: ' "${foo[@]}"; echo
:--foo: :--bar: :--baz='quux xyzzy':
```

It's passing `--baz='quux xyzzy'` as a single word, but it still includes the quote marks.

```console
$ foo=( --foo --bar ); foo+=( "--baz=quux xyzzy" )
$ printf ':%s: ' ${foo[*]}; echo
:--foo: :--bar: :--baz=quux: :xyzzy:
```

We got rid of the quote marks, but we're back to splitting `--baz=quux xyzzy`.

```console
$ printf ':%s: ' "${foo[@]}"; echo
:--foo: :--bar: :--baz=quux xyzzy:
```

This is what we want.

Does that make sense?

Comment (stevekuznetsov) [inline hack/util.sh]: Yes! What a nightmare.

**Comment by stevekuznetsov** [comment]
re[test] for fresh logs

**Comment by stevekuznetsov** [comment]
ShellCheck doesn't seem to want to install correctly... looking into it.

**Comment by stevekuznetsov** [comment]
@deads2k adding this linting steps has immediate benefits for us as it (in the way it is implemented here) stops any new Bash files from being checked in with poor code standards. I continue to think that tackling the largest, most difficult file for fixing all in one go is *not* the correct approach. I'm closing this PR, opening a new one with the linting step *only*. As that goes in we can move forward on making small changes over time to our codebase.

**Comment by deads2k** [comment]
> I continue to think that tackling the largest, most difficult file for fixing all in one go is not the correct approach.

Which file of reasonable complexity would you like to do?  e2e/core.sh maybe?  It reasonably representative.

**Comment by stevekuznetsov** [comment]
I'll do all of them, in some order, trying to make the review load on @Miciah low any given week. Huge files like `util` or`common` may take multiple passes.
